R/gg-plots.R

```diff
@@ -1292,6 +1292,108 @@ ggally_ratio <- function(
+#' @keywords hplot
+#' @export
+#' @examples
+#' data(tips, package = "reshape")
+#' ggally_count(tips, mapping = ggplot2::aes(x = smoker, y = sex))
+#'
+#' ggally_count(
+#'   as.data.frame(Titanic),
+#'   mapping = ggplot2::aes(x = Class, y = Survived, weight = Freq)
+#' )
+ggally_count <- function(data, mapping, ...) {
+
+  mapping <- mapping_color_to_fill(mapping)
+  vars <- unlist(lapply(mapping, mapping_string))
+  x_var <- vars["x"]
+  y_var <- vars["y"]
+  if (is.na(x_var)) stop("'x' aesthetic is required.")
+  if (is.na(y_var)) stop("'y' aesthetic is required.")
+
+  if (is.na(vars["weight"])) {
+    data$.weight <- 1
+  } else {
+    data$.weight <- data[[vars["weight"]]]
+  }
+
+  # compute formula for aggregation
+  # exclude weight and do not double count a var mapped to several aesthetics
+  f <- stats::as.formula(paste(
+    ".weight ~",
+    paste(unique(vars[names(vars) != "weight"]), collapse = "+")
+  ))
+
+  d <- stats::aggregate(f, data, sum, na.rm = TRUE)
+  names(d)[which(names(d) == ".weight")] <- ".n"
+
+  d$.x <- d[[x_var]] # keep original x &nd y
+  d$.y <- d[[y_var]]
+  d[[x_var]] <- as.numeric(d$.x)
+  # count total n for each combinaison of x and y
+  d$.n_xy <- stats::ave(d$.n, d$.x, d$.y, FUN = function(x) sum(x, na.rm = TRUE))
+  # proportion within a combinaison of x and y
+  d$.prop <- d$.n / d$.n_xy
+  d$.width <- sqrt(d$.n_xy) / max(sqrt(d$.n_xy)) * .9
+  d$.height <- d$.width * d$.prop
+  d$.cum_height <- stats::ave(d$.height, d$.x, d$.y, FUN = cumsum)
+  d[[y_var]] <- as.numeric(d$.y) + d$.cum_height - d$.height / 2 - d$.width / 2
+
+  # mapp height and width
+  # use aes_string() to avoid "no visible binding for global variable" NOTES
+  mapping$height <- aes_string(height = ".height")$height
+  mapping$width <- aes_string(width = ".width")$width
+
+  # remove weight aesthetic
+  mapping$weight <- NULL
+
+  # if fill corresponds to x or y, map to original factor
+  if (!is.na(vars["fill"]) & vars["fill"] == vars["x"])
+    mapping$fill <- aes_string(fill = ".x")$fill
+  else if (!is.na(vars["fill"]) & vars["fill"] == vars["y"])
+    mapping$fill <- aes_string(fill = ".y")$fill
+
+  p <- ggplot(data = d, mapping)
+
+  if ("fill" %in% names(list(...)) | !is.null(mapping$fill))
+    p <- p + geom_tile(...)
+  else
+    p <- p + geom_tile(fill = "grey50", ...)
+
+  p +
+    scale_x_continuous(
+      breaks = 1:length(levels(d$.x)),
+      labels = levels(d$.x)
+    ) +
+    scale_y_continuous(
+      breaks = 1:length(levels(d$.y)),
+      labels = levels(d$.y)
+    ) +
+    theme(panel.grid.minor = element_blank())
+}
+
+#' @rdname ggally_count
+#' @export
+#' @examples
+#'
+#' ggally_countDiag(tips, mapping = ggplot2::aes(x = smoker))
+#' ggally_countDiag(tips, mapping = ggplot2::aes(x = smoker, fill = sex))
+ggally_countDiag <- function(data, mapping, ...) {
+  data$.x_copy <- data[[mapping_string(mapping$x)]]
+  mapping$y <- aes_string(y = ".x_copy")$y
```

Comment:
Without setting the data, is it possible to copy the x mapping value?

```suggestion
  mapping$y <- mapping$x
```

Reply:
I tried but it's creating problem and you use a fill aesthetic in addition, as x and y position of the tiles will not be identical anymore.

Reply:
The other alternative would be, in `ggally_count()` instead of `ggally_countDiag`, to hard copy x and y and then to change the mapping towards the copy.

Reply:
Ah. I understand now about the hard coded value for `y`.

-----------------

True.  But I believe the x label would be incorrect if viewed.  I don't know of a good way to preserve the correct label, but hard coding the x/y names seems like a good idea for now.

Reply:
fix by https://github.com/ggobi/ggally/pull/321/commits/b9c71817857e713ffc4a8edf0de8ed2f35201860